malfeasance/handler.go

```diff
@@ -31,37 +29,51 @@ var (
 	errInvalidProof  = fmt.Errorf("%w: invalid proof", pubsub.ErrValidationReject)
 )
 
+type (
+	MalfeasanceHandlerV1 func(ctx context.Context, data scale.Type) (types.NodeID, []string, error)
```

Comment:
The `data scale.Type` is neither encoded or decoded in the handlers. It seems it only needs to be an _interface_ to support casting to the specific proof type. How about:
```suggestion
	MalfeasanceHandlerV1 func(ctx context.Context, data any) (types.NodeID, []string, error)
```

Reply:
It is already decoded by the `malfeasance` handler. I didn't change this, since it would require changing existing DTOs for malfeasance proofs from

```go
type Proof struct {
	// MultipleATXs | MultipleBallots | HareEquivocation | InvalidPostIndex
	Type uint8
	// AtxProof | BallotProof | HareProof | InvalidPostIndexProof
	Data scale.Type
}
```

to

```go
type Proof struct {
	Type uint8
	Data []byte
}
```

which I plan to do with V2 of malfeasance proofs.

The reason I use `scale.Type` and not `any` is because `any` allows passing anything, while `scale.Type` is more restrictive (i.e. only types with scale encoding can be passed). It's not ideal but I don't think I can be more more specific without breaking existing code 😕

Reply:
I understand, but using `scale.Type` is artificial in this context. I think this could be specific by using a dedicated interface with a dummy private method like:
```go
package wire

type Proof {
    isProof()
}

func (ap *AtxProof) isProof() {}
// other impls...
```

```go
package malfeasance

MalfeasanceHandlerV1 func(ctx context.Context, proof wire.Proof) (types.NodeID, []string, error)
```

As the method `isProof` is unexported, only types in the `malfeasance/wire` package can implement it. Looking from the point of view of the user of this API it becomes immediately obvious that he gets a proof, not something _encodable_.

Reply:
The only way to make this work is to have the `ProofData` interface (`Proof` is already a type) extend the `scale.Type` interface and put it in the `malfeasance/wire` package.